### ISSUE TITLE
Build broken on Python 3.10

### ISSUE THREAD

**Issue by njsmith** [MagicStack/immutables#46]
https://github.com/python/cpython/pull/20429

```
  immutables/_map.c: In function ‘map_node_bitmap_new’:
  immutables/_map.c:574:19: error: lvalue required as left operand of assignment
    574 |     Py_SIZE(node) = size;
        |                   ^
  immutables/_map.c: In function ‘map_node_collision_new’:
  immutables/_map.c:1359:19: error: lvalue required as left operand of assignment
   1359 |     Py_SIZE(node) = size;
        |                   ^
  error: command '/usr/bin/gcc' failed with exit code 1
```

@vstinner If the motivation is to make PyObject* opaque in the limited ABI, then why did you also make this change in the non-limited ABI?

**Comment by vstinner** [comment]
You should use Py_SET_SIZE() which is available since Python 3.9. To support Python 3.8 and older, you can use:

```
#ifndef Py_SET_SIZE
#  define Py_SET_SIZE(obj, size) do { Py_SIZE(obj) = (size); } while (0)
#endif
```

> @vstinner If the motivation is to make PyObject* opaque in the limited ABI, then why did you also make this change in the non-limited ABI?

My goal is to move slowly C extensions towards the limited C API, and so use the stable ABI.

For the rationale, see:

* https://mail.python.org/archives/list/capi-sig@python.org/thread/HYFQHLGNCRFBOMOOV2WUT46CQPXV5VQ7/
* https://bugs.python.org/issue39573
* https://github.com/vstinner/misc/blob/master/cpython/pep-opaque-c-api.rst

**Comment by vstinner** [comment]
Sorry, it's not a good idea to use `#ifndef Py_SET_SIZE`. Tomorrow, Py_SET_SIZE may become a function (I have a plan like that, especially for the limited C API). A better fix is:

```
#if PY_VERSION_HEX < 0x030900A4
#  define Py_SET_TYPE(obj, size) do { Py_TYPE(obj) = (size); } while (0)
#endif
```

I proposed https://github.com/python/cpython/pull/20610 to enhance the documentation to explain how to port existing code to Py_SET_SIZE(): to propose this macro for backward compatibility.

**Comment by vstinner** [comment]
I created PR #52 to add Python 3.10 support to immutables.

**Comment by vstinner** [comment]
I tested manually on Fedora 33 with Python 3.10.0a2:

```
$ python3.10 -m venv ENV
$ ENV/bin/python setup.py install
$ ENV/bin/python -m pip install pytest
$ ENV/bin/python -m pytest
```

It's now possible to build immutables on Python 3.10 and the test suite pass ("75 passed, 77 skipped in 4.45s").

**Comment by 1st1** [comment]
> 77 skipped

That's probably the C tests that were skipped. So there might be still some problem why the C extension wasn't built under 3.10

**Comment by vstinner** [comment]
> That's probably the C tests that were skipped. So there might be still some problem why the C extension wasn't built under 3.10

Oh. When tests are run from the source tree, "immutables._map" cannot be imported.

If I create a venv and run the venv Python outside the source tree, I can import the extension.

So the issue is just how I ran tests.

immutables are not installed by pip. So I'm not sure what is the safe way to install it and then run tests? I had to remove immutables/ directory from the source tree to keep tests and be able to import immutables._map.
